File: objectModel/Python/cdm/objectmodel/cdm_import_collection.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Optional, Union, TYPE_CHECKING

from cdm.enums import CdmObjectType

from .cdm_collection import CdmCollection

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusContext, CdmImport, CdmObject


class CdmImportCollection(CdmCollection):
    def __init__(self, ctx: 'CdmCorpusContext', owner: 'CdmObject'):
        super().__init__(ctx, owner, CdmObjectType.IMPORT)

    def append(self, obj: Union[str, 'CdmImport'], moniker: Optional[str] = None) -> 'CdmImport':
        if not isinstance(obj, str):
            if obj._previous_owner is not None:
                absolute_path = self.ctx.corpus.storage.create_absolute_corpus_path(obj.corpus_path, obj._previous_owner)
                    
                # Need to make the import path relative to the resolved manifest instead of the original manifest.
                obj.corpus_path = self.ctx.corpus.storage.create_relative_corpus_path(absolute_path, self.owner)

            return super().append(obj)

        import_value = super().append(obj)
        if moniker is not None:
            import_value.moniker = moniker

        return import_value

    def item(self, corpus_path: str, moniker: Optional[str] = None, check_moniker: Optional[bool] = True) -> 'CdmImport':
        for x in self:
            if check_moniker:
                if x.corpus_path == corpus_path and x.moniker == moniker:
                    return x
            else:
                if x.corpus_path == corpus_path:
                    return x
        return None

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/cdm/CdmImportCollection.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.cdm;

import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import java.util.List;

public class CdmImportCollection extends CdmCollection<CdmImport> {

  /**
   * Constructs a CdmImportCollection by using parent constructor.
   * and {@link CdmImport}.
   *
   * @param ctx   The context.
   * @param owner The owner of the collection.
   *              This class is customized for {@link CdmDocumentDefinition} owner.
   */
  public CdmImportCollection(final CdmCorpusContext ctx, final CdmObject owner) {
    super(ctx, owner, CdmObjectType.Import);
  }

  public CdmImport item(final String corpusPath, final String moniker, final boolean checkMoniker) {
    if (checkMoniker) {
      return this.allItems.stream().filter((x) -> corpusPath.equals(x.getCorpusPath())
              && ((x.getMoniker() == null && moniker == null) || (x.getMoniker() != null && x.getMoniker().equals(moniker)))).findFirst().orElse(null);
    } else {
      return this.allItems.stream().filter((x) -> corpusPath.equals(x.getCorpusPath())).findFirst().orElse(null);
    }
  }

  public CdmImport item(final String corpusPath, final String moniker) {
    return this.item(corpusPath, moniker, true);
  }

  @Override
  public CdmImport item(final String corpusPath) {
    return this.item(corpusPath, null);
  }

  @Override
  public CdmDocumentDefinition getOwner() {
    return (CdmDocumentDefinition) super.getOwner();
  }

  /**
   * Creates an import with the provided corpus path and adds it to the collection.
   * @param corpusPath The corpus path to be set for the import.
   * @return The created import which was added to the collection.
   */
  @Override
  public CdmImport add(final String corpusPath) {
    return super.add(corpusPath);
  }

  /**
   * Creates an import with the provided corpus path and adds it to the collection.
   * @param corpusPath The corpus path to be set for the import.
   * @param simpleRef Parameter is not used for this collection.
   *                  It is kept here for consistency with other {@link CdmCollection}-s.
   * @return The created import which was added to the collection.
   */
  @Override
  public CdmImport add(final String corpusPath, final boolean simpleRef) {
    final CdmImport cdmImport = this.getCtx()
        .getCorpus()
        .makeObject(this.getDefaultType(), corpusPath, simpleRef);
    this.add(cdmImport);
    return cdmImport;
  }

  /**
   * Creates an import with the provided corpus path
   * and provided moniker and adds it to the collection.
   * @param corpusPath The corpus path used to create the import.
   * @param moniker The moniker used to create the import.
   * @return The created import which was added to the collection.
   */
  public CdmImport add(final String corpusPath, final String moniker) {
    final CdmImport cdmImport = this.add(corpusPath);
    cdmImport.setMoniker(moniker);
    return cdmImport;
  }

  public void addAll(final List<CdmImport> importList) {
    importList.forEach(this::add);
  }
}
